forc-pkg/src/pkg.rs

```diff
@@ -1171,3 +1171,62 @@ fn test_source_git_pinned_parsing() {
         assert_eq!(&serialized, string);
     }
 }
+
+/// Format an error message for an absent `Forc.toml`.
+pub fn manifest_file_missing(curr_dir: PathBuf) -> anyhow::Error {
+    let message = format!(
+        "could not find `{}` in `{}` or any parent directory",
+        constants::MANIFEST_FILE_NAME,
+        curr_dir.display()
+    );
+    Error::msg(message)
+}
+
+/// Format an error message for failed parsing of a manifest.
+pub fn parsing_failed(project_name: &str, errors: Vec<CompileError>) -> anyhow::Error {
+    let error = errors
+        .iter()
+        .map(|e| e.to_friendly_error_string())
+        .collect::<Vec<String>>()
+        .join("\n");
+    let message = format!("Parsing {} failed: \n{}", project_name, error);
+    Error::msg(message)
+}
+
+/// Format an error message if an incorrect program type is present.
+pub fn wrong_program_type(
+    project_name: &str,
+    expected_type: TreeType,
+    parse_type: TreeType,
+) -> anyhow::Error {
+    let message = format!(
+        "{} is not a '{:?}' it is a '{:?}'",
+        project_name, expected_type, parse_type
+    );
+    Error::msg(message)
+}
+
+/// Format an error message if a given URL fails to produce a working node.
+pub fn fuel_core_not_running(node_url: &str) -> anyhow::Error {
+    let message = format!("could not get a response from node at the URL {}. Start a node with `fuel-core`. See https://github.com/FuelLabs/fuel-core#running for more information", node_url);
+    Error::msg(message)
+}
+
+/// Given the current directory and expected program type, determines whether the correct program type is present and returns a manifest.
+pub fn check_program_type(curr_dir: PathBuf, expected_type: TreeType) -> Result<Manifest> {
```

Comment:
I realise that for a couple of the upstream cases (`forc_deploy`, `forc_run`) that creating the `Manifest` inside this function saves a couple extra lines in those places. However, I think there's a bit of a conflation of responsibilities in having this function return a `Manifest`, and I'm not sure it's worth saving a couple lines downstream.

I still think passing in a reference to an existing `Manifest` is preferable as it keeps the separation of responsibilities clear. It also makes this function a little more modular for future use, e.g. if we come across a case where we've already loaded a `Manifest` but want to check the program type, we should not need to have to do more I/O and deserialize again just to check the program type. I hope this makes sense!

Reply:
Perhaps a clearer alternative might be:

```rust
impl Manifest {
    // ...

    //// The same as `Manifest::from_dir`, but checks the project is of the given program type before returning.
    pub fn from_dir_expecting_program_type(dir: &Path, ty: &TreeType) -> Result<Self> {
        // Implemented in terms of `Manifest::from_dir` and `check_program_type`.
    }
```

I think as a part of this, we could update the `Manifest::from_dir` function to automatically check parent directories in the case that the immediate directory doesn't contain a manifest. This might save us from always having to call `find_manifest_dir` everywhere before we can call `Manifest::from_dir`.

Reply:
>I think as a part of this, we could update the Manifest::from_dir function to automatically check parent directories in the case that the immediate directory doesn't contain a manifest. This might save us from always having to call find_manifest_dir everywhere before we can call Manifest::from_dir.
I do like that idea, would you mind if I went ahead and made these edits and then tackled this in another PR?

Reply:
Yeah that sounds fine to me!